File: api/send-digest.js
// api/send-digest.js - Отправка ежедневного дайджеста подписчикам
require('dotenv/config');
const { getTodayDigest } = require('../lib/digest');
const { getDigestSubscribers } = require('../lib/users');

const BOT_TOKEN = process.env.BOT_TOKEN;
const WEBAPP_URL = process.env.WEBAPP_URL || 'https://ispanskie-msk-bot-ver.vercel.app';

/**
 * Отправляет сообщение через Telegram Bot API
 */
async function sendMessage(chatId, text, replyMarkup = null) {
  const url = `https://api.telegram.org/bot${BOT_TOKEN}/sendMessage`;
  
  const body = {
    chat_id: chatId,
    text: text,
    parse_mode: 'HTML'
  };
  
  if (replyMarkup) {
    body.reply_markup = replyMarkup;
  }
  
  const response = await fetch(url, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body)
  });
  
  if (!response.ok) {
    const error = await response.text();
    throw new Error(`Telegram API error: ${response.status} - ${error}`);
  }
  
  return await response.json();
}

module.exports = async (req, res) => {
  // Безопасность: проверяем секретный ключ для cron
  const cronSecret = process.env.CRON_SECRET;
  const authHeader = req.headers['authorization'];
  
  if (cronSecret && authHeader !== `Bearer ${cronSecret}`) {
    console.log('[SEND_DIGEST] Unauthorized access attempt');
    return res.status(401).json({ success: false, error: 'Unauthorized' });
  }
  
  if (!BOT_TOKEN) {
    console.error('[SEND_DIGEST] BOT_TOKEN not configured');
    return res.status(503).json({ success: false, error: 'Bot not configured' });
  }
  
  console.log('[SEND_DIGEST] Starting digest send at', new Date().toISOString());
  
  try {
    // Получаем дайджест за сегодня
    const groqApiKey = process.env.GROQ_API_KEY || null;
    const digestData = await getTodayDigest(groqApiKey);
    
    if (!digestData || !digestData.digest) {
      console.log('[SEND_DIGEST] No digest data available');
      return res.json({ 
        success: true, 
        message: 'No digest to send',
        postsCount: 0,
        sentCount: 0
      });
    }
    
    const digestText = digestData.digest;
    const postsCount = digestData.postsCount || 0;
    
    console.log('[SEND_DIGEST] Digest generated:', {
      postsCount,
      textLength: digestText.length
    });
    
    // Получаем подписчиков
    const subscribers = await getDigestSubscribers();
    
    if (!subscribers || subscribers.length === 0) {
      console.log('[SEND_DIGEST] No subscribers found');
      return res.json({ 
        success: true, 
        message: 'No subscribers',
        postsCount,
        sentCount: 0
      });
    }
    
    console.log('[SEND_DIGEST] Found', subscribers.length, 'subscribers');
    
    // Кнопка "Открыть ХАБ" с прямой ссылкой на мини-приложение
    const keyboard = {
      inline_keyboard: [[
        { 
          text: '📱 Открыть ХАБ', 
          url: WEBAPP_URL
        }
      ]]
    };
    
    // Отправляем дайджест каждому подписчику
    let successCount = 0;
    let errorCount = 0;
    const errors = [];
    
    for (const user of subscribers) {
      try {
        const chatId = user.tgId;
        
        await sendMessage(chatId, digestText, keyboard);
        successCount++;
        
        console.log(`[SEND_DIGEST] Sent to user ${chatId}`);
        
        // Небольшая задержка, чтобы не превысить лимиты Telegram API
        if (subscribers.length > 1) {
          await new Promise(resolve => setTimeout(resolve, 50));
        }
      } catch (error) {
        errorCount++;
        const errorMsg = error.message || String(error);
        errors.push({ userId: user.tgId, error: errorMsg });
        console.error(`[SEND_DIGEST] Error sending to ${user.tgId}:`, errorMsg);
      }
    }
    
    const result = {
      success: true,
      postsCount,
      subscribersCount: subscribers.length,
      sentCount: successCount,
      errorCount,
      timestamp: new Date().toISOString()
    };
    
    if (errors.length > 0) {
      result.errors = errors;
    }
    
    console.log('[SEND_DIGEST] Completed:', result);
    
    return res.json(result);
  } catch (error) {
    console.error('[SEND_DIGEST] Critical error:', error);
    return res.status(500).json({ 
      success: false, 
      error: error.message || String(error),
      timestamp: new Date().toISOString()
    });
  }
};

File: send_digest.py
#!/usr/bin/env python3
"""
Скрипт для отправки ежедневного дайджеста подписчикам
Запускается по расписанию через cron
"""
import os
import sys
import json
import asyncio
import logging
from datetime import datetime
from telegram import Bot

logging.basicConfig(
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s',
    level=logging.INFO
)
logger = logging.getLogger(__name__)

USERS_FILE = 'users.json'

def load_users():
    """Загружает данные пользователей"""
    try:
        with open(USERS_FILE, 'r', encoding='utf-8') as f:
            return json.load(f)
    except FileNotFoundError:
        logger.error(f"Файл {USERS_FILE} не найден")
        return {'users': []}

def get_digest():
    """Получает дайджест через API"""
    import urllib.request
    import urllib.error
    
    api_url = os.getenv('API_URL', 'https://ispanskie-msk-bot-ver.vercel.app/api/digest')
    
    try:
        with urllib.request.urlopen(api_url) as response:
            data = json.loads(response.read().decode())
            if data.get('success') and data.get('digest'):
                return data['digest']
    except urllib.error.URLError as e:
        logger.error(f"Ошибка получения дайджеста: {e}")
    
    return None

async def send_digest_to_subscribers(bot_token: str):
    """Отправляет дайджест всем подписчикам"""
    # Загружаем подписчиков
    data = load_users()
    subscribers = [u for u in data['users'] if u.get('digestSubscription', False)]
    
    if not subscribers:
        logger.info("Нет подписчиков на дайджест")
        return
    
    # Получаем дайджест
    digest_data = get_digest()
    
    if not digest_data:
        logger.error("Не удалось получить дайджест")
        return
    
    digest_text = digest_data.get('digest', '')
    posts_count = digest_data.get('postsCount', 0)
    
    if not digest_text:
        logger.warning("Пустой дайджест")
        return
    
    # Инициализируем бота
    bot = Bot(token=bot_token)
    
    # Отправляем дайджест каждому подписчику
    success_count = 0
    error_count = 0
    
    for user in subscribers:
        try:
            tg_id = user['tgId']
            await bot.send_message(
                chat_id=tg_id,
                text=digest_text,
                parse_mode='HTML'
            )
            success_count += 1
            logger.info(f"Дайджест отправлен пользователю {tg_id}")
            
            # Небольшая задержка, чтобы не нарушать лимиты Telegram
            await asyncio.sleep(0.5)
        except Exception as e:
            error_count += 1
            logger.error(f"Ошибка отправки пользователю {user.get('tgId')}: {e}")
    
    logger.info(
        f"Дайджест отправлен: {success_count} успешно, {error_count} ошибок. "
        f"Новостей в дайджесте: {posts_count}"
    )

if __name__ == '__main__':
    TOKEN = os.getenv('TELEGRAM_BOT_TOKEN')
    
    if not TOKEN:
        logger.error("Не указан TELEGRAM_BOT_TOKEN в переменных окружения")
        sys.exit(1)
    
    logger.info(f"Запуск отправки дайджеста: {datetime.now()}")
    
    try:
        asyncio.run(send_digest_to_subscribers(TOKEN))
        logger.info("Отправка дайджеста завершена")
    except Exception as e:
        logger.error(f"Критическая ошибка: {e}")
        sys.exit(1)
